zipstream/ng.py

```diff
@@ -1317,5 +1402,40 @@ async def add_path(self, *args, **kwargs):
             return await to_thread(self._zip.add_path, *args, **kwargs)
 
         @functools.wraps(ZipStream.add)
-        async def add(self, data, *args, **kwargs):
-            return await to_thread(self._zip.add, data, *args, **kwargs)
+        async def add(
+            self,
+            data: AsyncIterable[bytes],
+            *args,
+            **kwargs,
+        ) -> None:
+            """
+            Add *data* to the zip archive where *data* is an asynchronous
+            iterable of ``bytes`` (e.g. an async generator that yields HTTP
+            chunks).
+
+            The method works by:
+
+            1. Creating a thread‑safe ``queue.Queue`` that is filled by a
+            background ``asyncio`` task running in the current event loop.
+            2. Passing a synchronous generator that reads from that queue
+            to the underlying ``ZipStream.add`` method (which runs in a
+            thread via ``to_thread``).
+
+            This keeps the event‑loop non‑blocking while the zip compression
+            happens in a separate thread.
+            """
+            # Turn the async iterable into a queue that the thread can read.
+            q = _queue_from_async_iterable(data)
+
+            # Build a *synchronous* generator that yields from the queue.
+            def sync_gen() -> Iterable[bytes]:
+                while True:
+                    chunk = q.get()  # blocks until a chunk is available
+                    if chunk is None:  # sentinel → end of stream
+                        break
+                    yield chunk
+
+            # Delegate to the original (thread‑based) ``add`` implementation.
+            # ``to_thread`` runs the whole call in a worker thread, so the
+            # zip compression stays off the event loop.
+            await asyncio.to_thread(self._zip.add, sync_gen(), *args, **kwargs)
```

Comment:
```suggestion
            await to_thread(self._zip.add, sync_gen(), *args, **kwargs)
```
I should have probably used this instead, right?